### PR TITLE
TSheets Quickbooks Time (Independent Publisher)

### PR DESCRIPTION
**Pull request by artesianst**
We have searched and there is no connector for the above mentioned product. We would like to develop and share this to Power Platform community.

### CODE REVIEW

**Comment by Yusadem131** [comment]
Hi, I am Yusuf, I am also working on Quickbook Connector. Can we collaborate together?

**Comment by Yusadem131** [comment]
This is my personal email(abdulakeemyusuf67@gmail.com) for further collaboration.

**Comment by artesianst** [comment]
> 

Hello, TSheets is a different product under Quickbooks. We are not working on Quickbooks Online or other products that are popularly known as Quickbooks.

**Comment by Yusadem131** [comment]
We may collaborate together on TSheets.

**Comment by artesianst** [comment]
> We may collaborate together on TSheets.

Thanks for letting us know. Let us wait for the approval of this connector and then we can have a meeting to take this forward.

**Comment by Yusadem131** [comment]
Okay. I will be looking forward to collaborating with you.

**Comment by Amjed-Ayoub** [comment]
Hello @artesianst,
Thank you so much for your new proposal submission. We are looking forward to seeing your connector artifacts soon.
Please let us know if you have any questions.
Thank you so much again for working with us.

**Comment by artesianst** [comment]
> Hello @artesianst, Thank you so much for your new proposal submission. We are looking forward to seeing your connector artifacts soon. Please let us know if you have any questions. Thank you so much again for working with us.

Hello @Amjed-Ayoub,

We are working to push the first iteration ASAP. Thanks for your encouraging words.

**Comment by artesianst** [comment]
> 

@microsoft-github-policy-service agree [company="Artesian Software Technologies LLP"]

**Comment by artesianst** [comment]
> @artesianst please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
> 1. **Definitions**.
>    **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>    approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>    **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>    Project, including but not limited to communication on electronic mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>    associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>    described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work, You may Submit them separately
>    to the Project if You (a) retain all copyright and license information that was in the materials as You
>    received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>    containing materials of a third party:” followed by the names of the third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>    guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your employer has intellectual
>    property rights in Your Submission by contract or applicable law, You must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
> 
> * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>   Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>   the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>   parties.
> * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>   Your patent claims that are necessarily infringed by the Submission or the combination of the
>   Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>   import or otherwise dispose of the Submission alone or with the Project.
> * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>   No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>   granted by implication, exhaustion, estoppel or otherwise.
> 
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>    have disclosed under Section 3). You represent that You have secured permission from Your employer to
>    make the Submission in cases where Your Submission is made in the course of Your work for Your
>    employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>    have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>    You later become aware that would make Your representations in this Agreement inaccurate in any
>    respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>    information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>     supersedes any and all prior agreements, understandings or communications, written or oral, between
>     the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

@microsoft-github-policy-service agree company="Artesian Software Technologies LLP"

**Comment by artesianst** [comment]
Hello @Amjed-Ayoub,

This is the first iteration that we are proposing. Please review and approve for production.

File: independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json:24]: @artesianst default responses is not allowed. Please define a proper response, please see [here](https://learn.microsoft.com/en-gb/connectors/custom-connectors/coding-standards#responses)

File: independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json:26]: @artesianst empty schema is not allowed please define a proper schema.

File: independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json:156]: @artesianst default responses is not allowed. Please define a proper response, please see [here](https://learn.microsoft.com/en-gb/connectors/custom-connectors/coding-standards#responses)

File: independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json:158]: @artesianst empty schema is not allowed please fix this in all your swagger file

File: independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json:230]: @artesianst default responses is not allowed. Please define a proper response, please see [here](https://learn.microsoft.com/en-gb/connectors/custom-connectors/coding-standards#responses)

File: independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json:390]: @artesianst default responses is not allowed. Please define a proper response, please see [here](https://learn.microsoft.com/en-gb/connectors/custom-connectors/coding-standards#responses)

File: independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json:546]: @artesianst default responses is not allowed. Please define a proper response, please see [here](https://learn.microsoft.com/en-gb/connectors/custom-connectors/coding-standards#responses)

File: independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json:232]: @artesianst empty schema is not allowed please fix this in all your swagger file

File: independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json:392]: @artesianst empty schema is not allowed please fix this in all your swagger file

File: independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/TSheets Quickbooks Time/apiDefinition.swagger.json:548]: @artesianst empty schema is not allowed please fix this in all your swagger file

File: independent-publisher-connectors/TSheets Quickbooks Time/apiProperties.json
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/TSheets Quickbooks Time/apiProperties.json:32]: @artesianst can you please confirm if you own this API or not? if not please change the stachOwner property to the name of the service owner.

**Review by Amjed-Ayoub — DISMISSED** [review]

**Comment by artesianst** [comment]
Please review @Amjed-Ayoub

**Comment by Amjed-Ayoub** [comment]
[[certify-connector]]

**Comment by Amjed-Ayoub** [comment]
Hello @artesianst,
I hope you are doing well.
Congratulations, your pull request is approved. We will move forward with the certification process. We are preparing your connector for production deployment and will queue up for the next deployment schedule. Once your connector onboards the next deployment schedule, it starts to deploy your connector in our production environments that typically takes 3-4 weeks. 
Please let us know if you have any questions.
Thank you very much for working with us.

**Comment by artesianst** [comment]
> Hello @artesianst, I hope you are doing well. Congratulations, your pull request is approved. We will move forward with the certification process. We are preparing your connector for production deployment and will queue up for the next deployment schedule. Once your connector onboards the next deployment schedule, it starts to deploy your connector in our production environments that typically takes 3-4 weeks. Please let us know if you have any questions. Thank you very much for working with us.

Hello @Amjed-Ayoub, thanks a lot for your quick response and guidance.

**Comment by Amjed-Ayoub** [comment]
Hello @artesianst ,
I hope you are doing well.
Your connector is in production deployment and queued up for the next deployment schedule. This process typically takes 3-4 weeks. Please note that the documentation for your connector will not be generated until the deployment process is in the final steps.
Please let us know if you have any questions.
Thank you very much for working with us.

**Comment by artesianst** [comment]
Thanks for the update @Amjed-Ayoub 🙏🏽

**Comment by artesianst** [comment]
> Hello @artesianst,
> 
> I hope you are doing well.
> 
> Unfortunately, we can't get a paid subscription for us in order to move forward with the deployment. If we use the 30 days trial version the connector will work for 30 days only and then it will break. The only option we have now is from you to give us a valid subscription to the API and create an account for us to move forward with the deployment process. Do you think this is something possible to do on your end?
> 
> Please let me know if you have any questions.
> 
> Thank you so much again for your patience and please accept our apologies for the inconvenience this may cause.
> 
> I look forward to hearing from you.

Hello @Amjed-Ayoub,

We also tried to get a Developer Account from TSheets to do the development but could not get it. Please try that route as it may be easier for you. We have our trial credentials that I can share with you but that may not be sufficient for you. Let me know how best to resolve this with Intuit TSheets.

Thanks

**Comment by Amjed-Ayoub** [comment]
Hi @artesianst,

Thank you so much for bringing this to my attention. I was able to create a developer account. we will try to deploy the connector with that account. Please expect to hear back from us within next week.
Thank you so much again for your patience and understanding.

**Comment by artesianst** [comment]
Thanks for the update @Amjed-Ayoub. We are eagerly waiting for the deployment.

**Comment by artesianst** [comment]
Hello @Amjed-Ayoub any update on this connector's production deployment? We are eagerly waiting.

**Comment by Amjed-Ayoub** [comment]
Hi @artesianst,

Apologies for the delay on this. We will start the deployment on Friday. if it succeeds, we will expedite to complete within next week. 

Thank you so much for your patience and understanding.

**Comment by artesianst** [comment]
Thanks for your quick reply @Amjed-Ayoub, we'll keep our fingers crossed :)

**Comment by artesianst** [comment]
Thanks for the update @Amjed-Ayoub

**Review by sriyen-msft — APPROVED** [review]

**Comment by sriyen-msft** [comment]
@artesianst Congratulations! The deployment of your connector and documentation page is now complete, We have a short [survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbR0BHMrjhL0hDmckROosW6AFURERIVUUzMk5XRzJGV1dJM1Q5RVZJUjFXSC4u) on the connector certification process that we would like for you to complete. Your feedback is appreciated and will help us improve the program, Thanks.

**Comment by Ajinder31** [comment]
Thank you for the update @sriyen-msft. 

When we tried to use the QuickBooks time connector in PA we got the following error. 

TSheets_Connector_Error

Please let us know if we're missing anything. I just clicked sign in button in connector and it throws this error.

**Comment by Amjed-Ayoub** [comment]
Hello @artesianst, thank you so much for the confirmation. we will investigate this issue within next week.
We expect that the Auth app for your connector is not accepting the client id from the developer account that we created. if this is the case then we will need a valid subscription for QuickBooks.
I will confirm within next week. 
Thank you so much again,
Best regards
Amjed Ayoub

**Comment by Amjed-Ayoub** [comment]
Hello @artesianst I hope you are doing well.

We were able to fix the connection issue with your connector in one of our test regions.
Web capture_27-12-2022_16756_make test powerautomate com

We need your help with fixing this issue by submitting a new update for your connector and raise a new pull request. 
Please replace the current apiProperties.json file with the below code.
```
{
  "properties": {
    "connectionParameters": {
      "token": {
        "type": "oauthSetting",
        "oAuthSettings": {
          "identityProvider": "oauth2",
          "clientId": "Enter your client ID",
          "scopes": [],
          "redirectMode": "Global",
          "redirectUrl": "https://global.consent.azure-apim.net/redirect",
          "properties": {
            "IsFirstParty": "False",
            "IsOnbehalfofLoginSupported": false
          },
          "customParameters": {
            "authorizationUrl": {
              "value": "https://rest.tsheets.com/api/v1/authorize"
            },
            "tokenUrl": {
              "value": "https://rest.tsheets.com/api/v1/grant"
            },
            "refreshUrl": {
              "value": "https://rest.tsheets.com/api/v1/grant"
            }
          }
        }
      },
      "token:clientId": {
        "type": "string",
        "uiDefinition": {
          "constraints": {
            "required": "true",
            "hidden": "false"
          },
          "description": "Please enter the client Id.",
          "displayName": "Client Id - Your QuickBooks clientId"
        }
      },
      "token:clientSecret": {
        "type": "securestring",
        "uiDefinition": {
          "constraints": {
            "required": "true",
            "hidden": "false"
          },
          "description": "Please enter the client secret.",
          "displayName": "Client Secret"
        }
      }
    },
    "iconBrandColor": "#da3b01",
    "capabilities": [
      "actions"
    ],
    "publisher": "Artesian Software Technologies LLP",
    "stackOwner": "TSheets QuickBooks Time"
  }
}
```

Please let me know if you have any questions.

Thank you so much

**Comment by artesianst** [comment]
Hello @Amjed-Ayoub, we have holidays here till Monday. We'll take care of it as soon as we resume office. Thanks for your help and understanding.

**Pull request merged**
